### PR TITLE
Adding notebook workspaces API to Cosmos Swagger

### PR DESCRIPTION
**Pull request by shurd**
### Latest improvements:
<i>MSFT employees can try out our new experience at <b>[OpenAPI Hub](https://aka.ms/openapiportal) </b> - one location for using our validation tools and finding your workflow. 
</i><br>
### Contribution checklist:
- [x] I have reviewed the [documentation](https://github.com/Azure/adx-documentation-pr/wiki/Overall-basic-flow) for the workflow.
- [x] [Validation tools](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-checklist.md#validation-tools-for-swagger-checklist) were run on swagger spec(s) and have all been fixed in this PR.
- [x] The [OpenAPI Hub](https://aka.ms/openapiportal) was used for checking validation status and next steps.
### ARM API Review Checklist
- [x] Service team MUST add the "WaitForARMFeedback" label if the management plane API changes fall into one of the below categories. 
- adding/removing APIs.
- adding/removing properties.
- adding/removing API-version. 
- adding a new service in Azure.

Failure to comply may result in delays for manifest application. Note this does not apply to data plane APIs.
- [ ] If you are blocked on ARM review and want to get the PR merged urgently, please get the ARM oncall for reviews (RP Manifest Approvers team under Azure Resource Manager service) from IcM and reach out to them. 
Please follow the link to find more details on [API review process](https://armwiki.azurewebsites.net/rp_onboarding/ResourceProviderOnboardingAPIRevieworkflow.html).

### CODE REVIEW

File: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/stable/2019-08-01/notebook.json
Comment (wmengmsft) [inline specification/cosmos-db/resource-manager/Microsoft.DocumentDB/stable/2019-08-01/notebook.json]: How do we indicate in Swagger that only the "default" notebook workspace name is supported?

Comment (shurd) [reply specification/cosmos-db/resource-manager/Microsoft.DocumentDB/stable/2019-08-01/notebook.json]: I originally thought that it had to be parameterized for us to not be flagged for "swagger completeness", but I think I was incorrect. I will remove this parameter, force the path to be "/notebookWorkspaces/default", and update examples accordingly.

File: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/stable/2019-08-01/notebook.json
Comment (KrisBash) [inline specification/cosmos-db/resource-manager/Microsoft.DocumentDB/stable/2019-08-01/notebook.json]: For all API responses in this spec, please add a "default" response to capture error cases. It should have a schema consistent with the [common error contract](https://github.com/Azure/azure-resource-manager-rpc/blob/master/v1.0/common-api-details.md#error-response-content). 
Here is an example:
https://github.com/Azure/azure-rest-api-specs/blob/0cf70260108cb40bc68d408674c187ed065f8c64/specification/batch/resource-manager/Microsoft.Batch/stable/2019-08-01/BatchManagement.json#L105

File: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/stable/2019-08-01/notebook.json
Comment (KrisBash) [inline specification/cosmos-db/resource-manager/Microsoft.DocumentDB/stable/2019-08-01/notebook.json]: "list" is a preferred prefix for POST actions that return connection strings/secrets. Actions with "list" as the prefix verb can be called from an ARM template

**Review by KrisBash — CHANGES_REQUESTED** [review]
A few comments

File: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/stable/2019-08-01/notebook.json
Comment (wmengmsft) [reply specification/cosmos-db/resource-manager/Microsoft.DocumentDB/stable/2019-08-01/notebook.json]: This is for future-proofing. Overtime it is possible to have additional user provided parameters here.
Another reason we want to model this as a separate resource is that it represents a set of functionalities that are very distinct from those provided by the parent resource. We want to avoid bloating the parent resource too much. This resource also has its own set of actions that we want to keep separate from the parent to avoid confusion.

File: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/stable/2019-08-01/notebook.json
Comment (wmengmsft) [reply specification/cosmos-db/resource-manager/Microsoft.DocumentDB/stable/2019-08-01/notebook.json]: Thanks for the feedback. Are you saying actions MUST have a "list" prefix to be usable in ARM template? If so, we will make this change.

Comment (KrisBash) [reply specification/cosmos-db/resource-manager/Microsoft.DocumentDB/stable/2019-08-01/notebook.json]: Yep

File: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/stable/2019-08-01/notebook.json
Comment (MyronFanQiu) [inline specification/cosmos-db/resource-manager/Microsoft.DocumentDB/stable/2019-08-01/notebook.json]: The commonly used word is `parameters`. I don't think `createParameters` is a good naming. Is it able to change it to a meaningful word like `notebookParameter`?

File: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/stable/2019-08-01/notebook.json
Comment (MyronFanQiu) [inline specification/cosmos-db/resource-manager/Microsoft.DocumentDB/stable/2019-08-01/notebook.json]: Does it mean `provisioningState`? Just a sharing, there are some logic rely on the word `provisioningState` in the runtime of SDKs. If the `status` is the same thing with `provisioningState`, it's better to use the word `provisioningState`.

**Comment by MyronFanQiu** [comment]
@shurd Hello. Could you address ARM's comments?

**Comment by shurd** [comment]
Hi @MyronFanQiu we are planning on making the naming change the ARM team suggested, but will not be able to deploy it until at least January, so we are going to wait to merge these changes until then. I will address your comments once we get closer to having the change in production.

**Comment by MyronFanQiu** [comment]
@shurd I will remove the label once you finish this PR. 😄

File: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/stable/2019-08-01/notebook.json
Comment (sanjaiganesh) [inline specification/cosmos-db/resource-manager/Microsoft.DocumentDB/stable/2019-08-01/notebook.json]: >					 [](start = 0, length = 5)

change tabs-> 4 spaces

**Comment by sanjaiganesh** [comment]
@shurd please address the comments.

File: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/stable/2019-08-01/notebook.json
Comment (shurd) [reply specification/cosmos-db/resource-manager/Microsoft.DocumentDB/stable/2019-08-01/notebook.json]: This is different than provisioning state.

File: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/stable/2019-08-01/notebook.json
Comment (shurd) [reply specification/cosmos-db/resource-manager/Microsoft.DocumentDB/stable/2019-08-01/notebook.json]: I think it fixed this last time I ran prettier.

File: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/stable/2019-08-01/notebook.json
Comment (shurd) [reply specification/cosmos-db/resource-manager/Microsoft.DocumentDB/stable/2019-08-01/notebook.json]: The change has been deployed.

**Comment by shurd** [comment]
@MyronFanQiu @sanjaiganesh We deployed the changes to our service, and I have made the requested changes. Can you please re-review?

**Comment by shurd** [comment]
@MyronFanQiu @KrisBash We made the suggested changes in our RP/swagger. Can you please re-review?

**Review by mmyyrroonn — APPROVED** [review]
LGTM. Sorry for the delay due to Chinese Sprint festival.

**Comment by wmengmsft** [comment]
LGTM

**Comment by shurd** [comment]
@anthony-c-martin can you help with ARM signoff?

File: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/stable/2019-08-01/notebook.json
Comment (pilor) [inline specification/cosmos-db/resource-manager/Microsoft.DocumentDB/stable/2019-08-01/notebook.json]: >Create [](start = 43, length = 6)

CreateOrUpdate to follow standard conventions

File: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/stable/2019-08-01/notebook.json
Comment (pilor) [inline specification/cosmos-db/resource-manager/Microsoft.DocumentDB/stable/2019-08-01/notebook.json]: >"202" [](start = 10, length = 5)

this operation needs the x-ms-long-running-operation flag if it is async

File: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/stable/2019-08-01/notebook.json
Comment (pilor) [inline specification/cosmos-db/resource-manager/Microsoft.DocumentDB/stable/2019-08-01/notebook.json]: >202 [](start = 11, length = 3)

needs x-ms-long-running-operation if it is async

File: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/stable/2019-08-01/notebook.json
Comment (pilor) [inline specification/cosmos-db/resource-manager/Microsoft.DocumentDB/stable/2019-08-01/notebook.json]: >status [](start = 9, length = 6)

this should be provisioningState and have valid terminal states.  This is required for your PUT operation to be handled async correctly.  [https://github.com/Azure/azure-resource-manager-rpc/blob/master/v1.0/Addendum.md#creatingupdating-using-put](https://github.com/Azure/azure-resource-manager-rpc/blob/master/v1.0/Addendum.md#creatingupdating-using-put)

File: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/stable/2019-08-01/notebook.json
Comment (pilor) [inline specification/cosmos-db/resource-manager/Microsoft.DocumentDB/stable/2019-08-01/notebook.json]: please flag these as readonly for completeness

File: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/stable/2019-08-01/notebook.json
Comment (pilor) [inline specification/cosmos-db/resource-manager/Microsoft.DocumentDB/stable/2019-08-01/notebook.json]: >"notebookServerEndpoint" [](start = 8, length = 24)

is there a reason to replicate this on the resource as well as this connectionInfo result?

**Review by pilor — CHANGES_REQUESTED** [review]
few comments above

File: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/stable/2019-08-01/notebook.json
Comment (wmengmsft) [reply specification/cosmos-db/resource-manager/Microsoft.DocumentDB/stable/2019-08-01/notebook.json]: Hi @pilor , my understanding is that provisioningState is only required for tracked resources and resources following 201 + provisioningState pattern. 
These APIs here are proxy APIs. We also use 202 + Location header async pattern for these APIs, which based on our testing are working correctly. Since most of our other proxy resources already work this way, can we continue to following the same pattern?

File: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/stable/2019-08-01/notebook.json
Comment (wmengmsft) [reply specification/cosmos-db/resource-manager/Microsoft.DocumentDB/stable/2019-08-01/notebook.json]: We felt the notebook server endpoint property is relevant in both context so it is added as a convenience for the user. If you strongly feel it is redundant, we can eliminate one of them.

**Comment by MyronFanQiu** [comment]
@shurd Hello. Could you solve ARM's comments?

**Comment by MyronFanQiu** [comment]
@shurd Kindly ping

**Comment by MyronFanQiu** [comment]
@pilor Could you sign off if author has resolved all your comments?

File: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/stable/2019-08-01/notebook.json
Comment (pilor) [reply specification/cosmos-db/resource-manager/Microsoft.DocumentDB/stable/2019-08-01/notebook.json]: The async pattern is the same for both proxy and tracked resources.  I don't see any other instance of using a "status" property to represent the provisioning status of a resource in cosmos DB swaggers.  If a status property is needed you should be using the normal 201 + provisioningState pattern.  If not, 202 + location header is "ok" but not the preferred approach

Comment (MyronFanQiu) [reply specification/cosmos-db/resource-manager/Microsoft.DocumentDB/stable/2019-08-01/notebook.json]: @shurd Kindly ping

Comment (shurd) [reply specification/cosmos-db/resource-manager/Microsoft.DocumentDB/stable/2019-08-01/notebook.json]: We are working on making this change to our RP. I will update once we have resolved it.

**Review by pilor — APPROVED** [review]

**Comment by shurd** [comment]
@MyronFanQiu We have resolved the last comments. Can you please help merge this change?

File: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/stable/2019-08-01/notebook.json
Comment (MyronFanQiu) [reply specification/cosmos-db/resource-manager/Microsoft.DocumentDB/stable/2019-08-01/notebook.json]: @shurd If I understand correctly, there will be another PR to change it. right? Is this one resolved? @pilor

**Review by mmyyrroonn — APPROVED** [review]

**Comment by MyronFanQiu** [comment]
@KrisBash Hi. Could you approve this PR if you don't have any more concern?

**Comment by MyronFanQiu** [comment]
@shurd Regarding to the `status` and the `provisioningState` comment, the swagger spec would be changed in another PR after the fix of server finished. Is it correct?

**Comment by shurd** [comment]
@MyronFanQiu, as I understand it, @wmengmsft and @pilor discussed the `status` change offline and determined that we can keep it as-is. There will be no additional change.

**Comment by MyronFanQiu** [comment]
> @MyronFanQiu, as I understand it, @wmengmsft and @pilor discussed the `status` change offline and determined that we can keep it as-is. There will be no additional change.

Good to know. Wait for @KrisBash's approval.

**Comment by MyronFanQiu** [comment]
@akning-ms Hi. I think @KrisBash and @pilor are both from ARM's team. Do you think it's ok to force merge this one since ARM has signed off? @KrisBash marked a request change for this PR.

**Comment by wmengmsft** [comment]
@MyronFanQiu @KrisBash, all requested change have been resolved. Can we please proceed with merging this PR?

**Pull request merged**
